### PR TITLE
202410 async notebook reworks

### PR DESCRIPTION
**Pull request by petermarshallio**
This change includes a new notebook focused on historical data that includes pagination and calls to the API directly rather than via the package. It also includes major edits to the existing "all timeline" notebook, focusing it on real-time ingestion, and shortening it considerably.

### CODE REVIEW

**Review by hevansDev — CHANGES_REQUESTED** [review]

File: notebooks/03-query/21-query-async-historical.ipynb [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "472589e4-1026-4b3b-bb79-eedabb2b44c4",
+   "metadata": {},
+   "source": [
+    "## Create a table using batch ingestion\n",
+    "\n",
+    "<!-- Use these cells if you are using batch ingestion for your notebook. -->\n",
+    "\n",
+    "Run the following cell to create a table using batch ingestion. Notice {the use of X as a timestamp | only required columns are ingested | WHERE / expressions / GROUP BY are front-loaded | partitions on X period and clusters by Y}.\n",
```
[the remark this answers is not in the document]
Comment (petermarshallio) [reply notebooks/03-query/21-query-async-historical.ipynb:133]: Oops this is where I've left the boilerplate text in from the template.

File: notebooks/03-query/21-query-async-historical.ipynb [hunk 2]
```diff
@@ -0,0 +1,610 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d88e9033-f322-4c62-ab05-78db3dea5c45",
+   "metadata": {},
+   "source": [
+    "Run the cell above until some segments are shown without a server and tier.\n",
+    "\n",
+    "Notice that, when you run the following query, results are limited to data fetched to a historical."
```
[the remark this answers is not in the document]
Comment (petermarshallio) [reply notebooks/03-query/21-query-async-historical.ipynb:293]: Agreed.

**Review by hevansDev — APPROVED** [review]

**Comment by techdocsmith** [comment]
Still need to review #21, but many of the comments that applied to 14 will apply here.

File: notebooks/03-query/21-query-async-historical.ipynb [hunk 3]
```diff
@@ -0,0 +1,610 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "3a92ffc2-856c-420f-b058-51042c7e6be0",
+   "metadata": {},
+   "source": [
+    "## Apply retention rules\n",
+    "\n",
+    "Force some of the data to be left on deep storage by changing the retention load rules for a table.\n",
+    "\n",
+    "In the next cell, the retention rules ensure that Druid picks up data after 11th November 2005 to load to Historicals in the `_default_tier` tier. All other data is caught by the second rule, which ensures no data is cached on Historicals.\n",
```
Comment (techdocsmith) [inline notebooks/03-query/21-query-async-historical.ipynb:223]:
```suggestion
    "The next cell defines retention rules the table. The first configures Druid to load data after 11th November 2005 onto Historicals in the `_default_tier` tier. The second rule, which applies to all remaining data, prevents Druid from loading data onto Historicals.\n",
```

File: notebooks/03-query/21-query-async-historical.ipynb [hunk 4]
```diff
@@ -0,0 +1,610 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d88e9033-f322-4c62-ab05-78db3dea5c45",
+   "metadata": {},
+   "source": [
+    "Run the cell above until some segments are shown without a server and tier.\n",
+    "\n",
+    "Notice that, when you run the following query, the results exclusively consist of data that has been pre-fetched to a historical."
```
Comment (techdocsmith) [inline notebooks/03-query/21-query-async-historical.ipynb:293]:
```suggestion
    "Notice that the results from the following query consist only of data that has been pre-cached to a Historical."
```

File: notebooks/03-query/21-query-async-historical.ipynb [hunk 5]
```diff
@@ -0,0 +1,610 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "045df49c-24ad-4eff-9b5f-0a30bfd47bf1",
+   "metadata": {},
+   "source": [
+    "### Call the API using the druid_api package\n",
+    "\n",
+    "The `async_sql` method handles the [necessary steps](https://druid.apache.org/docs/latest/tutorials/tutorial-query-deep-storage#query-from-deep-storage) not only to submit the query, but to retrieve the results.\n",
```
Comment (techdocsmith) [inline notebooks/03-query/21-query-async-historical.ipynb:330]:
```suggestion
    "The `async_sql` method of the Druid API library handles the [necessary steps](https://druid.apache.org/docs/latest/tutorials/tutorial-query-deep-storage#query-from-deep-storage) to both submit the query and retrieve the results.\n",
```

File: notebooks/03-query/21-query-async-historical.ipynb [hunk 6]
```diff
@@ -0,0 +1,610 @@
...
+    "\n",
+    "Run the following cell, which uses the `async_sql` method of the `druid_api` package to call the API and return the results."
```
Comment (techdocsmith) [inline notebooks/03-query/21-query-async-historical.ipynb:332]:
```suggestion
    "Run the following cell to run an ansynchronous query and return the results."
```
I think line 330 is clear enough about the `async_sql` method so we don't need to repeat it here.

File: notebooks/03-query/21-query-async-historical.ipynb [hunk 7]
```diff
@@ -0,0 +1,610 @@
...
+    "  COUNT(*) as \"events\"\n",
+    "FROM \"{table_name}\"\n",
+    "WHERE TIME_IN_INTERVAL(\"__time\",'2005-11-16/P1W')\n",
+    "GROUP BY 1\n",
+    "'''\n",
+    "\n",
+    "result = sql_client.async_sql(sql)\n",
+    "result.wait_until_done()\n",
+    "\n",
+    "print(json.dumps(result.rows, indent=2))"
```
Comment (techdocsmith) [inline notebooks/03-query/21-query-async-historical.ipynb:354]: Why can't we do table formatted results?

File: notebooks/03-query/21-query-async-historical.ipynb [hunk 8]
```diff
@@ -0,0 +1,610 @@
...
+   "cell_type": "markdown",
+   "id": "4b2d3341-37a8-4eb1-80f4-6391b55b6b42",
+   "metadata": {},
+   "source": [
+    "### Call the API directly\n",
+    "\n",
+    "Run the following cell to submit the query directly to the API.\n",
+    "\n",
+    "* The `request_json` object contains the query and the necessary context parameters.\n",
+    "* The API is called using `requests.post` with the results stored in `response`.\n",
```
Comment (techdocsmith) [inline notebooks/03-query/21-query-async-historical.ipynb:375]:
```suggestion
    "* The API call uses Python's `requests.post` method with the results stored in the `response`.\n",
```

File: notebooks/03-query/21-query-async-historical.ipynb [hunk 9]
```diff
@@ -0,0 +1,610 @@
...
+    "* A pretty print is made of the JSON response from the API."
```
Comment (techdocsmith) [inline notebooks/03-query/21-query-async-historical.ipynb:376]:
```suggestion
    "* Pretty print nicely formats the JSON response from the API."
```

File: notebooks/03-query/21-query-async-historical.ipynb [hunk 10]
```diff
@@ -0,0 +1,610 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c2371778-e6e8-43ab-8bad-d808d7e6e28d",
+   "metadata": {},
+   "source": [
+    "Once the status shows as \"SUCCESS\", retrieve the result by running the following cell.\n",
+    "\n",
+    "The URL is constructed from:\n",
```
Comment (techdocsmith) [inline notebooks/03-query/21-query-async-historical.ipynb:431]:
```suggestion
    "The URL comprises:\n",
```

File: notebooks/03-query/21-query-async-historical.ipynb [hunk 11]
```diff
@@ -0,0 +1,610 @@
...
+    "* The URL to the API, appended with the query Id.\n",
+    "* The page of the results to return.\n",
+    "* The format of the results to return.\n",
```
Comment (techdocsmith) [inline notebooks/03-query/21-query-async-historical.ipynb:434]:
```suggestion
    "* The result format.\n",
```

File: notebooks/03-query/21-query-async-historical.ipynb [hunk 12]
```diff
@@ -0,0 +1,610 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "8ac7a4b5-c467-4304-8f06-d3eb5c5bf165",
+   "metadata": {},
+   "source": [
+    "Notice the `queryId` in the response object.\n",
+    "\n",
+    "Run the next cell to use the API to see the status of the query as it executes. The cell captures the `queryId` and uses this to construct a GET to the API."
```
Comment (techdocsmith) [inline notebooks/03-query/21-query-async-historical.ipynb:406]:
```suggestion
    "Run the next cell to monitor query status during execution. The cell captures the `queryId` to construct a GET request for the API."
```

File: notebooks/03-query/21-query-async-historical.ipynb [hunk 13]
```diff
@@ -0,0 +1,610 @@
...
+    "\n",
+    "print(json.dumps(job_result_json, indent=2))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "c2371778-e6e8-43ab-8bad-d808d7e6e28d",
+   "metadata": {},
+   "source": [
+    "Once the status shows as \"SUCCESS\", retrieve the result by running the following cell.\n",
```
Comment (techdocsmith) [inline notebooks/03-query/21-query-async-historical.ipynb:429]:
```suggestion
    "Once Druid returns a \"SUCCESS\" status, you can run the following cell to retrieve the results.\n",
```

File: notebooks/03-query/21-query-async-historical.ipynb [hunk 14]
```diff
@@ -0,0 +1,610 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "8988cb6f-8e51-4cf9-bdd6-047272ca5dae",
+   "metadata": {},
+   "source": [
+    "## Paginate results\n",
+    "\n",
+    "Use the `rowsPerPage` query context parameter to control the size of the results page.\n",
+    "\n",
+    "In the `query_request_json`, the page size is set by providing a parameter in the call to the Python API, and the destination of the query has been set to [`durableStorage`](https://druid.apache.org/docs/latest/operations/durable-storage).\n",
```
Comment (techdocsmith) [inline notebooks/03-query/21-query-async-historical.ipynb:464]:
```suggestion
    "Note that the `query_request_json` sets the page size limits the number of results per page to 3 rows and sets the destination of the query to [`durableStorage`](https://druid.apache.org/docs/latest/operations/durable-storage).\n",
```

File: notebooks/03-query/21-query-async-historical.ipynb [hunk 15]
```diff
@@ -0,0 +1,610 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "54b8d5fe-ba85-4b5b-9669-0dd47dfbccd1",
+   "metadata": {},
+   "source": [
+    "## Summary\n",
+    "\n",
+    "* SELECTs can run online (interactive API) or offline (asynchronous API).\n",
```
Comment (techdocsmith) [inline notebooks/03-query/21-query-async-historical.ipynb:570]:
```suggestion
    "* SELECTs can run interactively or asynchronously.\n",
```

File: notebooks/03-query/21-query-async-historical.ipynb [hunk 16]
```diff
@@ -0,0 +1,610 @@
...
+    "* The offline API can access data that has not been prefetched to Historicals.\n",
```
Comment (techdocsmith) [inline notebooks/03-query/21-query-async-historical.ipynb:571]:
```suggestion
    "* The asynchronous API can access data in deep storage that has not been pre-cached to Historicals.\n",
```

File: notebooks/03-query/21-query-async-historical.ipynb [hunk 17]
```diff
@@ -0,0 +1,610 @@
...
+    "* Durable Storage can be used to hold results of queries.\n",
```
Comment (techdocsmith) [inline notebooks/03-query/21-query-async-historical.ipynb:572]:
```suggestion
    "* You can send the results of queries to Durable Storage.\n",
```

File: notebooks/03-query/21-query-async-historical.ipynb [hunk 18]
```diff
@@ -0,0 +1,610 @@
...
+    "* Results can be paginated and can have different formats.\n",
```
Comment (techdocsmith) [inline notebooks/03-query/21-query-async-historical.ipynb:573]:
```suggestion
    "* You can use various formats for query results and you can paginate the results.\n",
```

**Review by techdocsmith — CHANGES_REQUESTED** [review]
Many of the same comments for Notebook 14 apply to Notebook 21

**Review by techdocsmith — CHANGES_REQUESTED** [review]
Many of the same comments for Notebook 14 apply to Notebook 21

File: notebooks/03-query/21-query-async-historical.ipynb (hunk 7 above)
Comment (petermarshallio) [reply notebooks/03-query/21-query-async-historical.ipynb:335]: I couldn't find the appropriate method to do this when using the async client.

**Pull request merged**